coderd/prebuilds/reconcile.go

```diff
@@ -0,0 +1,110 @@
+package prebuilds
+
+import (
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/util/slice"
+)
+
+type ActionType int
+
+const (
+	ActionTypeBackoff ActionType = iota
+	ActionTypeCreate
+	ActionTypeDelete
+)
+
+// GlobalSnapshot represents a full point-in-time snapshot of state relating to prebuilds across all templates.
+type GlobalSnapshot struct {
+	Presets             []database.GetTemplatePresetsWithPrebuildsRow
+	RunningPrebuilds    []database.GetRunningPrebuiltWorkspacesRow
+	PrebuildsInProgress []database.CountInProgressPrebuildsRow
+	Backoffs            []database.GetPresetsBackoffRow
+}
+
+// PresetSnapshot is a subset of GlobalSnapshot but specifically for a single preset.
+type PresetSnapshot struct {
+	Preset     database.GetTemplatePresetsWithPrebuildsRow
+	Running    []database.GetRunningPrebuiltWorkspacesRow
+	InProgress []database.CountInProgressPrebuildsRow
+	Backoff    *database.GetPresetsBackoffRow
+}
+
+// ReconciliationActions represents a single action needed to reconcile the current state with the desired state.
+// Exactly one field will be set based on the ActionType.
+type ReconciliationActions struct {
+	// ActionType determines which field is set and what action should be taken
+	ActionType ActionType
+
+	// Create is set when ActionType is ActionTypeCreate and indicates the number of prebuilds to create
+	Create int32
+
+	// DeleteIDs is set when ActionType is ActionTypeDelete and contains the IDs of prebuilds to delete
+	DeleteIDs []uuid.UUID
+
+	// BackoffUntil is set when ActionType is ActionTypeBackoff and indicates when to retry creating prebuilds
+	BackoffUntil time.Time
+}
+
```

Comment:
- Add comment here. Emphasize it's derived from `PresetSnapshot` and ReconciliationActions is calculated based on `ReconciliationState `.
- Diff between `PresetSnapshot` & `ReconciliationState`?
- Update comments for PresetSnapshot?